src/signal.cr

```diff
@@ -123,6 +124,34 @@ enum Signal : Int32
   end
 
   @@setup_default_handlers = Atomic::Flag.new
+  @@setup_segfault_handler = Atomic::Flag.new
+  @@segfault_handler = LibC::SigactionHandlerT.new { |sig, info, data|
+    # Capture fault signals (SEGV, BUS) and finish the process printing a backtrace first
+
+    # Determine if the SEGV was inside or 'near' the top of the stack
+    # to check for potential stack overflow. 'Near' is a small
+    # amount larger than a typical stack frame, 4096 bytes here.
+    addr = info.value.si_addr
+
+    is_stack_overflow =
+      begin
+        stack_top = Pointer(Void).new(Fiber.current.@stack.address - 4096)
+        stack_bottom = Fiber.current.@stack_bottom
+        stack_top <= addr < stack_bottom
+      rescue e
+        Crystal::System.print_error "Error while trying to determine if a stack overflow has occurred. Probable memory corruption\n"
+        false
+      end
+
+    if is_stack_overflow
+      Crystal::System.print_error "Stack overflow (e.g., infinite or very deep recursion)\n"
+    else
+      Crystal::System.print_error "Invalid memory access (signal %d) at address 0x%lx\n", sig, addr
```

Comment:
Not for this PR, but would it be helpful to print the name of the signal?  In my experience the number is hard to understand.  Perhaps something like:

```suggestion
      Crystal::System.print_error "Invalid memory access (signal %d: SIG%s) at address 0x%lx\n", sig, Signal.new(sig), addr
```